# Introduction to R

## An Introduction to R: From Installation to Reading and Writing Data.

### R and RStudio

<img src="Logos/R_logo.png" style="float: right; margin-right: 15px;" width="70"> 

**R is a free, open-source programming language and software environment specifically designed for statistical computing, data visualization, and data science.** It can be used through many Graphic User Interfaces (GUI) my preference is to use `RStudio` and `Jupyter Notebook` but `VSCode` is good and you can also code in `base R` (as well as many others).

<div style="clear: both;"></div>

**RStudio is the premier Integrated Development Environment (IDE) for the R programming language, developed by [Posit](https://posit.co).** It provides a (GUI) for working with R, making it easier to use for those who are less familiar with command-line interfaces.

### Installations

- Download R from the official site: https://cran.r-project.org

- Download RStudio Desktop from: https://posit.co/download/rstudio-desktop/

### Basics

In this tutorial, we will rely on using code written in RStudio and locations of things (**Script**, **Console**, **Environment**, **Plots/Help**) will be RStudio specific but the code could be run in any GUI.

### R projects

Creating a new project in RStuido is a very simple task. To create a project:

1. Click `Project: (None)` on the top right of the RStudio window then `New Project`.
2. Select `New Directory` and then `New Project`.
3. Name the project into any project name of your choice.
4. Save it i a location of your choice (make sure you understand your computers file system and where the files are saved).
5. Your RStudio should now show the name of your project where there used to be `Project: (None)`.

### RStudio panels

RStudio has four panes each occupying a quadrant of your screen: **Source Editor**, **Console**, **Workspace Browser** (and History), and **Plots** (and Files, Packages, Help). These can also be adjusted under the ‘Preferences’ menu.

#### Source Editor

The source editor is located at the top left pane of RStudio. It can help you edit and execute lines of code. Source editor in RStudio has the advantage of providing syntax highlighting, code completion, and smart indentation. You can see different colours for numbers and functions, there is also highlighting to help you count brackets. We execute R code directly from the source editor by having the cursor on that line in the script or selecting multiple lines and then clicking the **run**  button or `command+enter` (mac) or `ctrl+enter` (windows).

```
The hash (#) tells R not to run any of the text on that line to the right of the symbol. This is the standard way of commenting R code or writing notes on our script 
```

#### The console

The console is located at the bottom left of pane of RStudio. This is where you can type code that executes immediately. This is also known as the command line. Entering code in the command line is intuitive and easy. For example, we can use R to do simple calculations with values.

### Simple Mathematics

In [1]:
2 + 2

[1] 4

In [2]:
5*6/4

[1] 7.5

In [3]:
2^3

[1] 8

In [4]:
(5+4)/3 # Mathematical order matters

[1] 3

In [8]:
sqrt(121)

[1] 11

In [9]:
pi

[1] 3.141593

R evaluates expressions from right to left, following standard mathematical rules.

We can also use either `<-` or `=` to assigna value, list or dataframe into an **object** so we can save it to R's global environment (top right pane) for use later. An **object** is something that is saved in temporary memory on the global environment.

In [10]:
x <- 5
y <- 10
x + y

[1] 15

In [11]:
krill_density <- 12.4
adder_records <- 58

krill_density

[1] 12.4

Object names should be clear and descriptive. Avoid naming something 'a' unless you truly mean something temporary.

### Functions

In R, we can use functions to do tasks for us. They normally precede a parenthesis (), some are named after what they do and some are less well named. Within functions, we have 'arguments', which are like options basically. What you put into these arguments will define how the perform. 

One commonly used function is `c()`, which concatenates (combines) elements together into a vector. **A vector is a series of values.**

In [12]:
numbers <- c(1, 2, 3, 4, 5)

We can apply functions to an object

In [13]:
mean(numbers)

[1] 3

If we want to check the documentation for a package we can go to the Help window, or type ? before the function name.

In [ ]:
?mean()

We can then perform different functions between objects

In [14]:
numbers * 2
numbers + 10

[1]  2  4  6  8 10

[1] 11 12 13 14 15

In [16]:
mean(numbers * krill_density)

[1] 37.2

We can even save the results to a new object

In [17]:
krill_number_density <- numbers * krill_density

Then we can look at what is in the new object by running the object or printing it (`print()`)

In [18]:
krill_number_density

[1] 12.4 24.8 37.2 49.6 62.0

In [19]:
print(krill_number_density)

[1] 12.4 24.8 37.2 49.6 62.0


We can also create data systematically with R

For example a sequence of 10 values going up by 1

In [20]:
Sequence <- seq(from = 1, to = 10,by = 0.5)

Sequence

[1]  1.0  1.5  2.0  2.5  3.0  3.5  4.0  4.5  5.0  5.5  6.0  6.5  7.0  7.5  8.0
[16]  8.5  9.0  9.5 10.0

In [21]:
Another_sequence <- c(1:200)

Another_sequence

[1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
[109] 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
[127] 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
[145] 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
[163] 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
[181] 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
[199] 199 200

### **Data types**

In R there are various types of data, the most common four are **Numeric**, **Integer**, **Character** and **Factor**.

**Logical** and **Comples** are also data types however they are very rarely used explicitly.

- Numeric data is any real numbers so 8 or 12.3 or 1.00000002 etc.
- Integer data is just whole numbers 3, 4, 111 etc.
- Character data are words or letters surrounded by quotations (either ” or ‘) such as “A”, “Red”, ’Treated’, Character data has no order to it in Rs ‘mind’
- Factor data is like character data but r (or you) have assigned an order to it e.g. “A”, “B”, “C”.

In [37]:
numeric_vector <- c(1.3,5.8,5.122,3.00,7.12)

# we change between data types with these functions
integer_vector <- as.integer(c(1, 5, 5, 3, 7)) 

character_vector <- c("krill", "adder", "fynbos")

logical_vector <- c(TRUE, FALSE, TRUE)

# Notice how r automatically orders alphabetically if we don't tell it the order
factor_vector <- as.factor(c("This","is","A","Character", "Vector"))

Check types:

In [38]:
class(numeric_vector)
class(integer_vector)
class(character_vector)
class(logical_vector)
class(factor_vector)

[1] "numeric"

[1] "integer"

[1] "character"

[1] "logical"

[1] "factor"

Mixing types forces coercion:

In [39]:
c(1, "a")

[1] "1" "a"

R will convert everything to character. It tries to maintain consistency within a vector.

We can also change between types (even if they don’t fit that description)

In [40]:
Convert_numbers_to_characters <- as.character(numeric_vector)

Convert_numbers_to_characters

[1] "1.3"   "5.8"   "5.122" "3"     "7.12"

Now our numbers are thought of as characters, so we can’t apply numeric operations to them!

### Matrix

A matrix is a multi-dimensional data structure that can hold elements of the same data type (it could be Numeric, Character, Integer, Factor). A matrix can be created using the `matrix()` function in R.

In [41]:
Matrix_numeric <- as.matrix(c(1.3,5.8,5.122,3.00,7.12))

Matrix_character <- as.matrix(c("This","is","A","Character Matrix"))

### DataFrame

A dataframe is the most commonly used 'container' for data in R. A dataframe is a combination of multiple groups of data types (it could be Numeric, Character, Integer, Factor), with two dimensions. Each element of a dataframe must have the same length as the other element.

In [30]:
ecology_data <- data.frame(
  species = c("krill", "adder", "proteaceae"),
  abundance = c(120, 15, 45),
  site = c("A", "B", "C")
)

View it

In [31]:
ecology_data

species,abundance,site
<chr>,<dbl>,<chr>
krill,120,A
adder,15,B
proteaceae,45,C


Access columns

In [32]:
ecology_data$species
ecology_data$abundance

[1] "krill"      "adder"      "proteaceae"

[1] 120  15  45

Check structure

In [33]:
str(ecology_data)
summary(ecology_data)

'data.frame':	3 obs. of  3 variables:
 $ species  : chr  "krill" "adder" "proteaceae"
 $ abundance: num  120 15 45
 $ site     : chr  "A" "B" "C"


   species            abundance         site          
 Length:3           Min.   : 15.0   Length:3          
 Class :character   1st Qu.: 30.0   Class :character  
 Mode  :character   Median : 45.0   Mode  :character  
                    Mean   : 60.0                     
                    3rd Qu.: 82.5                     
                    Max.   :120.0                     

### Lists

Lists are multiple groups of a combination of data types or objects (it could be Numeric, Character, Integer, Factor or vectors, dataframes or matrices of these), with two dimensions. Each element in a list can be a different length to the other elements. Lists are useful for storing complex data structures, can also be accessed using indexing.

In [42]:
List_numeric <- list(c(1.3, 5.8, 5.122, 3.00, 7.12),
                   c(1, 5, 5, 3, 7))

List_from_vectors <- list(character_vector, Matrix_numeric, Matrix_numeric)

List_different_lengths <- list(Item1 = c(1, 2, 3, 4, 5, 6), Item2 = c("a", "B", "C", "D"), Item3 = seq(from = 1, to = 100, by = 1))

List_different_lengths

$Item1
[1] 1 2 3 4 5 6

$Item2
[1] "a" "B" "C" "D"

$Item3
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

### Array

An array refers to multi-dimensional data structures that can hold elements of the same data type. It can have any number of dimensions and its dimensions can be named. **Multi-dimensional arrays are common in modelling and spatial data.**

In [35]:
Array_1d <- array(c(Matrix_numeric,c(1.3,5.8,5.122,3.00,7.12)),dim=c(5))
Array_2d <- array(c(Matrix_numeric,c(1.3,5.8,5.122,3.00,7.12)),dim=c(5,2))
Array_3d <- array(c(Matrix_numeric,c(1.3,5.8,5.122,3.00,7.12)),dim=c(5,2,2))

### **Indexing**

Object have dimensions and there is a technique we can use called **indexing** to select elements of an object. We often use square brackets to do this. If the object is 1 dimensional, one number will return one value.

In [43]:
numeric_vector[4]

[1] 3

If the object is 2 dimensional, one number will return that column

In [45]:
ecology_data[2]

abundance
<dbl>
120
15
45


By adding a comma we can select the from both dimensions (rows first, then columns)

In [47]:
ecology_data[3, 2]

[1] 45

If we want all rows but only a specific column we add a comma without a number

In [48]:
ecology_data[, 2]

[1] 120  15  45

And vice verse

In [49]:
ecology_data[3,]

,species,abundance,site
,<chr>,<dbl>,<chr>
3,proteaceae,45,C


### Packages

R relies on packages, groups of specific functions, which can be installed from the internet and loaded into a script. Most common functions used in R are contained within the base package. Packages are collections of code called functions that automate complex mathematical, statistical tasks. From Base R we can use the `install.packages()` function to install a package from online repositories. R automatically assumes you want to download packages from CRAN (the official online repository) but sometimes you might want to download from other repositories.

In [ ]:
# install.packages("tidyverse")

You only have to do this when you first want the package or want to update it or when you have updated R. Once a package is installed we have to tell R that we want to use functions from this package so we load it:

In [51]:
library(tidyverse)

This needs to be run every new R session when this package is used.

If you are working in ecology or GIS, useful packages include:
- `sf` for spatial data
- `terra` for raster data
- `vegan` for community ecology
- `ggplot2` for visualisation

You can search CRAN here: https://cran.r-project.org/web/packages/

### **Loading data**

Most real work starts with importing data.

In [ ]:
data <- read.csv("your_file.csv")

Always check:

In [ ]:
head(data)
str(data)
summary(data)

Never analyse data you have not inspected.

### Data inspection

First we can make some data into a dataframe, explore this data, then save it as a file and then read the file back into R. R has some very useful random and non-random data generation functions

In [52]:
Year <- seq(from = 1950, to = 2023, by = 1)
Treatment <- c("Control", "Treatment 1", "Treatment 2")
Rep <- seq(from = 1, to = 10, by = 1)

These are three vectors, which we can check information about them with a few simple functions

In [53]:
length(Year)

[1] 74

In [54]:
summary(Year)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1950    1968    1986    1986    2005    2023 

In [55]:
length(Treatment)

[1] 3

In [56]:
summary(Treatment)

   Length     Class      Mode 
        3 character character 

In [58]:
length(Rep)

[1] 10

In [59]:
summary(Rep)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.00    3.25    5.50    5.50    7.75   10.00 

We want to combine these vectors so we have a row for each **Rep**, **Year** and **Treatment**, we can do this by expanding the grid and create a new dataframe called **df**. We can inspect specific elements of a dataframe too.

In [62]:
df <- expand.grid(Year = Year, Treatment = Treatment, Rep = Rep)

class(df) # Type of object

[1] "data.frame"

In [63]:
nrow(df) # Number of rows

[1] 2220

In [64]:
ncol(df) # Number of columns

[1] 3

In [65]:
dim(df) # Dimensions of object

[1] 2220    3

In [66]:
head(df) # The first 6 rows of the df

,Year,Treatment,Rep
,<dbl>,<fct>,<dbl>
1,1950,Control,1
2,1951,Control,1
3,1952,Control,1
4,1953,Control,1
5,1954,Control,1
6,1955,Control,1


In [67]:
tail(df) # The last 6 rows of the df

,Year,Treatment,Rep
,<dbl>,<fct>,<dbl>
2215,2018,Treatment 2,10
2216,2019,Treatment 2,10
2217,2020,Treatment 2,10
2218,2021,Treatment 2,10
2219,2022,Treatment 2,10
2220,2023,Treatment 2,10


This **df** is all the metadata we want for our dataframe that we want to now make up some response data

In [68]:
# We need the response to be same length as the df so we use nrow() for the number of values we want.
Response <- rnorm(n = nrow(df), mean = 15, sd = 8) 

We can then combine this to our **df**, the dollar sign ($) is used to select one dimension (column) from within an object (here a dataframe). The column Response isn’t present in the data but by assigning our Response vector to it with `df$Response <- Response` it adds a new column called Response to the dataframe df.

In [70]:
df$Response <- Response

### **Data saving**

We now can save the **df** we created with the Resoponce column to a new folder using the `write.csv()` function from base R, or even better the `write_csv()` function from the `readr` package.

To save to our folder we only need to say the directory we want the file saved to, followed by a '/', then the name of the file with file extension. Inside reading and writing functions such as `write_csv()` the main argument will be where is the file to be saved to or taken from and we write this out as a character string inside quotations.

In [ ]:
write_csv(df,"NewFolderName/OurNewFile.csv")

### **Getting help**

R has built-in documentation.

In [ ]:
?mean
help(lm)

If you are stuck:

- Stack Overflow: https://stackoverflow.com
- RStudio Cheatsheets: https://posit.co/resources/cheatsheets/

Learning R is less about memorising commands and more about understanding patterns.

#### Where to next?

Once you understand objects, vectors, and data frames, you are ready for:

- Data wrangling
- Plotting with ggplot2
- Spatial data handling
- Statistical modelling

Those topics are covered in the next sections of Les Hyperion.